src/main/java/com/alvarium/hash/HashProvider.java

```diff
@@ -0,0 +1,24 @@
+package com.alvarium.hash;
+
+import com.alvarium.hash.none.NoneProvider;
+import com.alvarium.hash.types.HashType;
+
+public interface HashProvider {
+  // derive converts data to an hash value
+  String derive(byte[] data);
+
+  public static HashProvider factory(HashType type) throws UnsupportedHashTypeException {
```

Comment:
I would not recommend this. Consider that any class implementing the HashProvider interface will itself have a factory() method on it if treated as its interface. This would imply some reason to use an MD5 HashProvider to instantiate a SHA256 HashProvider, in essence pushing an additional responsibility onto the HashProvider classes.

The factory should be implemented separately from the classes returned by the factory.

Reply:
Me and Ali were debating on whether or not to include it into the the interface. Your point of view makes complete sense to the argument. I will make sure to communicate this with Ali and the team and will act accordingly.

Reply:
fixed in 614ad8b71c75928fc926ceaed04342e82e57f65f commit.